packages/cli/src/databases/repositories/execution.repository.ts

```diff
@@ -399,21 +405,89 @@ export class ExecutionRepository extends Repository<ExecutionEntity> {
 
 		if (!executions.length) {
 			if (deleteConditions.ids) {
-				LoggerProxy.error('Failed to delete an execution due to insufficient permissions', {
+				Logger.error('Failed to delete an execution due to insufficient permissions', {
 					executionIds: deleteConditions.ids,
 				});
 			}
 			return;
 		}
 
 		const executionIds = executions.map(({ id }) => id);
-		const binaryDataManager = BinaryDataManager.getInstance();
-		await binaryDataManager.deleteBinaryDataByExecutionIds(executionIds);
-
 		do {
 			// Delete in batches to avoid "SQLITE_ERROR: Expression tree is too large (maximum depth 1000)" error
-			const batch = executionIds.splice(0, 500);
-			await this.delete(batch);
+			const batch = executionIds.splice(0, this.deletionBatchSize);
+			await this.softDelete(batch);
 		} while (executionIds.length > 0);
 	}
+
+	async pruneBySoftDeleting() {
+		Logger.verbose('Pruning (soft-deleting) execution data from database');
+
+		const maxAge = config.getEnv('executions.pruneDataMaxAge'); // in h
+		const maxCount = config.getEnv('executions.pruneDataMaxCount');
+
+		// Find ids of all executions that were stopped longer that pruneDataMaxAge ago
+		const date = new Date();
+		date.setHours(date.getHours() - maxAge);
+
+		const toPrune: Array<FindOptionsWhere<IExecutionFlattedDb>> = [
+			// date reformatting needed - see https://github.com/typeorm/typeorm/issues/2286
+			{ stoppedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)) },
+		];
+
+		if (maxCount > 0) {
+			const executions = await this.find({
+				select: ['id'],
+				skip: maxCount,
+				take: 1,
+				order: { id: 'DESC' },
+			});
+
+			if (executions[0]) {
+				toPrune.push({ id: LessThanOrEqual(executions[0].id) });
+			}
+		}
+
+		const executionIds = (
+			await this.find({
+				select: ['id'],
+				where: toPrune,
+				take: this.deletionBatchSize,
+			})
+		).map(({ id }) => id);
+		await this.softDelete(executionIds);
```

Comment:
We can actually replace this code with a single update, where we set `deletedAt = now() where stoppedAt <= "{cutDate}"` - would be much more performant.